diff -Naur firewall-init-2.99.5/firewall.d/functions firewall-init-2.99.5-p/firewall.d/functions
--- firewall-init-2.99.5/firewall.d/functions	Sun May 20 22:39:48 2001
+++ firewall-init-2.99.5-p/firewall.d/functions	Wed May  8 16:22:58 2002
@@ -135,6 +135,7 @@
 	$iptables -P OUTPUT ACCEPT
 	$iptables -P FORWARD DROP
 
+	$iptables -A INPUT -i lo -j ACCEPT
 	$iptables -A INPUT -m state --state ESTABLISHED,RELATED -j ACCEPT
 }
 
